File: frontend/src/singleton/SessionManager.js
class SessionManager {
  static instance = null
  constructor() {
    if (SessionManager.instance) {
      throw new Error(
        'Ya existe una instancia de SessionManager. Usa getInstance()'
      )
    }
    this.token = null
    this.usuario = null
    console.log('SessionManager — instancia única creada')
  }

  static getInstance() {
    if (SessionManager.instance === null) {
      SessionManager.instance = new SessionManager()
    }
    return SessionManager.instance
  }

  guardarSesion(token, usuario) {
    this.token = token
    this.usuario = usuario
    localStorage.setItem('token', token)
    localStorage.setItem('usuario', JSON.stringify(usuario))
    console.log('Sesión guardada para:', usuario.email)
  }

  cerrarSesion() {
    console.log('Sesión cerrada para:', this.usuario?.email)
    this.token = null
    this.usuario = null
    localStorage.removeItem('token')
    localStorage.removeItem('usuario')
  }

  getToken() {
    return this.token || localStorage.getItem('token')
  }

  getUsuario() {
    return this.usuario || JSON.parse(localStorage.getItem('usuario'))
  }

  getSesionActiva() {
    return this.getToken() !== null
  }
}

export default SessionManager

File: ms-usuarios/src/main/java/com/smartlogix/ms_usuarios/service/SessionManager.java
package com.smartlogix.ms_usuarios.service;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;


public class SessionManager {

    private static SessionManager instance = null;
    private final Map<String, String> sesionesActivas = new ConcurrentHashMap<>();

    private SessionManager() {
        System.out.println("SessionManager inicializado — instancia única creada");
    }

    public static SessionManager getInstance() {
        if (instance == null) {
            synchronized (SessionManager.class) {
                if (instance == null) {
                    instance = new SessionManager();
                }
            }
        }
        return instance;
    }

    public void registrarSesion(String token, String email) {
        sesionesActivas.put(token, email);
        System.out.println("Sesión registrada para: " + email);
    }

    public void cerrarSesion(String token) {
        String email = sesionesActivas.remove(token);
        System.out.println("Sesión cerrada para: " + email);
    }

    public boolean sesionActiva(String token) {
        return sesionesActivas.containsKey(token);
    }

    public String getEmailPorToken(String token) {
        return sesionesActivas.get(token);
    }

    public int totalSesionesActivas() {
        return sesionesActivas.size();
    }
}